x-pack/plugins/endpoint/public/applications/endpoint/store/test_utils.ts

```diff
@@ -0,0 +1,138 @@
+   * especially when run in a CI environment.
+   *
+   * @param actionType
+   */
+  waitForAction: <T extends keyof ActionsMap<A>>(actionType: T) => Promise<ActionsMap<A>[T]>;
+  /**
+   * A property holding the information around the calls that were processed by the  internal
+   * `actionSpyMiddlware`. This property holds the information typically found in Jets's mocked
+   * function `mock` property - [see here for more information](https://jestjs.io/docs/en/mock-functions#mock-property)
+   *
+   * **Note**: this property will only be set **after* the `actionSpyMiddlware` has been
+   * initialized (ex. via `createStore()`. Attempting to reference this property before that time
+   * will throw an error.
+   * Also - do not hold on to references to this property value if `jest.clearAllMocks()` or
+   * `jest.resetAllMocks()` is called between usages of the value.
+   */
+  dispatchSpy: jest.Mock<Dispatch<A>>['mock'];
+  /**
+   * Redux middleware that enables spying on the action that are dispatched through the store
+   */
+  actionSpyMiddleware: ReturnType<MiddlewareFactory<S>>;
+}
+
+/**
+ * Creates a new instance of middleware action helpers
+ * Note: in most cases (testing concern specific middleware) this function should be given
+ * the state type definition, else, the global state will be used.
+ *
+ * @example
+ * // Use in Policy List middleware testing
+ * const middlewareSpyUtils = createSpyMiddleware<PolicyListState>();
+ * store = createStore(
+ *    policyListReducer,
+ *    applyMiddleware(
+ *      policyListMiddlewareFactory(fakeCoreStart, depsStart),
+ *      middlewareSpyUtils.actionSpyMiddleware
+ *    )
+ * );
+ * // Reference `dispatchSpy` ONLY after creating the store that includes `actionSpyMiddleware`
+ * const { waitForAction, dispatchSpy } = middlewareSpyUtils;
+ * //
+ * // later in test
+ * //
+ * it('...', async () => {
+ *   //...
+ *   await waitForAction('serverReturnedPolicyListData');
+ *   // do assertions
+ *   // or check how action was called
+ *   expect(dispatchSpy.calls.length).toBe(2)
+ * });
+ */
+export const createSpyMiddleware = <
+  S = GlobalState,
+  A extends AppAction = AppAction
+>(): MiddlewareActionSpyHelper<S, A> => {
+  type ActionWatcher = (action: A) => void;
+
+  const watchers = new Set<ActionWatcher>();
+  let spyDispatch: jest.Mock<Dispatch<A>>;
+
+  return {
+    waitForAction: async actionType => {
+      // Error is defined here so that we get a better stack trace that points to the test from where it was used
+      const err = new Error(`action '${actionType}' was not dispatched within the allocated time`);
+
+      return new Promise<ActionsMap<A>[typeof actionType]>((resolve, reject) => {
+        const watch: ActionWatcher = action => {
+          if (action.type === actionType) {
+            watchers.delete(watch);
+            clearTimeout(timeout);
+            resolve((action as unknown) as ActionsMap<A>[typeof actionType]);
```

Comment:
I had to this this casting here (which I think is safe being that we are inside of the `if()` clause) because I kept getting errors in TS around the types not matching. It might have something to do with the fact that the `ActionWatcher` is defined to take in `AppAction` (`A`), but the return value is actually from the `ActionMap`. They should match up (in my opinion), but they are not 😞